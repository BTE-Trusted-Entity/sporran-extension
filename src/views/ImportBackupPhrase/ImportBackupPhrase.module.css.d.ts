export const __esModule: true;
export const container: string;
export const heading: string;
export const info: string;
export const items: string;
export const item: string;
export const input: string;
export const wordErrorTooltip: string;
export const tooltip: string;
export const buttonLine: string;
export const button: string;
export const errorTooltip: string;
