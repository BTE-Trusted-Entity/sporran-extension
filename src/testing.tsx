export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
