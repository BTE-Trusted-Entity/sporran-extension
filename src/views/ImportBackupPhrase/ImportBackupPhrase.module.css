.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.info {
  composes: subline from '../../components/Typography/Typography.module.css';
  text-align: center;
  margin: 0.5rem 0;
}

.items {
  height: calc(264rem / 16);
  width: calc(284rem / 16);
  list-style-type: decimal-leading-zero;
  list-style-position: inside;
  padding-left: 0;
  display: flex;
  flex-flow: column wrap;
  text-align: left;
}

.item::marker {
  /* composes: smallImportant from '../../components/Typography/Typography.module.css'; */
  font-family: 'Montserrat Medium', sans-serif;
  font-size: 0.875rem;
  color: var(--color-orange);
}

.item {
  composes: input from '../../components/Input/Input.module.css';
  box-sizing: border-box;
  width: 8.5rem;
  margin-bottom: 0.75rem;
  padding: 0.375rem 1rem;
  position: relative;
}

.input {
  composes: inputText from '../../components/Typography/Typography.module.css';
  position: relative;
  top: -1.625rem;
  left: -1rem;
  width: 4.5rem;
  height: 2rem;
  padding-top: 0;
  border: 0 solid transparent;
  border-left-width: 3rem;
  border-right-width: 1rem;
  background: none;
}

.wordErrorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  z-index: 1;
}

.input:not(:focus) + .tooltip {
  display: inline;
}

.buttonLine {
  position: relative;
}

.button {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
  width: calc(210rem / 16);
  display: block;
  margin: 0 auto;
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
}
