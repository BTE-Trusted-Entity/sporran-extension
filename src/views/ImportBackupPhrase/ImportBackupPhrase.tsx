import {
  ChangeEvent,
  ChangeEventHandler,
  ClipboardEvent,
  ClipboardEventHandler,
  FormEvent,
  useCallback,
  useState,
} from 'react';
import DEFAULT_WORDLIST from '@polkadot/util-crypto/mnemonic/bip39-en';
import { mnemonicValidate } from '@polkadot/util-crypto';
import { Utils } from '@kiltprotocol/sdk-js';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './ImportBackupPhrase.module.css';

import { LinkBack } from '../../components/LinkBack/LinkBack';

type BackupPhrase = string[];

const noError = null;

function isAllowed(word: string) {
  return DEFAULT_WORDLIST.includes(word);
}

function isInvalid(
  backupPhrase: BackupPhrase,
  expectedAddress?: string,
): string | null {
  const t = browser.i18n.getMessage;

  const mnemonic = backupPhrase.join(' ');
  if (!mnemonicValidate(mnemonic)) {
    return t('view_ImportBackupPhrase_error_invalid_backup_phrase');
  }

  const { address } = Utils.Crypto.makeKeypairFromUri(mnemonic, 'sr25519');

  const noNeedToCompare = !expectedAddress;
  const matchesExpectations = expectedAddress === address;
  if (noNeedToCompare || matchesExpectations) {
    return noError;
  }

  return t('view_ImportBackupPhrase_error_mismatched_backup_phrase');
}

function isMalformed(backupPhrase: BackupPhrase): string | null {
  const length = backupPhrase.filter(Boolean).length;
  const hasNoWords = length === 0;
  const hasAllWords = length === 12;
  const allIsFine = hasAllWords || hasNoWords;
  if (allIsFine) {
    return noError;
  }
  const t = browser.i18n.getMessage;
  return t('view_ImportBackupPhrase_error_backup_phrase_length');
}

interface WordInputProps {
  word: string;
  index: number;
  handleInput: ChangeEventHandler<HTMLInputElement>;
  handlePaste?: ClipboardEventHandler<HTMLInputElement>;
}

function WordInput({
  word,
  index,
  handleInput,
  handlePaste,
}: WordInputProps): JSX.Element {
  const t = browser.i18n.getMessage;

  const hasError = Boolean(word && !isAllowed(word));

  return (
    <li className={styles.item}>
      <input
        aria-label={(index + 1).toString()}
        id={index.toString()}
        name={index.toString()}
        className={styles.input}
        type="text"
        onInput={handleInput}
        onPaste={handlePaste}
        value={word}
        autoFocus={index === 0}
      />
      <output
        htmlFor={index.toString()}
        className={styles.wordErrorTooltip}
        hidden={!hasError}
      >
        {t('view_ImportBackupPhrase_error_typo')}
      </output>
    </li>
  );
}

interface Props {
  type?: 'import' | 'reset';
  address?: string;
  onImport: (backupPhrase: string) => void;
}

export function ImportBackupPhrase({
  type = 'import',
  address,
  onImport,
}: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const [backupPhrase, setBackupPhrase] = useState<BackupPhrase>(
    Array(12).fill(''),
  );
  const modified = backupPhrase.join('') !== '';

  const error =
    modified &&
    [
      isMalformed(backupPhrase),
      isInvalid(backupPhrase, type === 'reset' ? address : undefined),
    ].filter(Boolean)[0];

  const handleInput = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const { name, value } = event.target;
      backupPhrase[Number(name)] = value.trim().toLowerCase();
      setBackupPhrase([...backupPhrase]);
    },
    [backupPhrase],
  );

  const handlePaste = useCallback((event: ClipboardEvent) => {
    const text = event.clipboardData.getData('text');
    if (!text) {
      return;
    }
    const values = text.trim().split(/\s+/);
    if (values.length !== 12) {
      return;
    }
    setBackupPhrase(values);
    event.preventDefault();
  }, []);

  const handleSubmit = useCallback(
    (event: FormEvent) => {
      event.preventDefault();
      if (error) {
        return;
      }
      const backupPhraseString = backupPhrase.join(' ');
      onImport(backupPhraseString);
    },
    [backupPhrase, error, onImport],
  );

  return (
    <section className={styles.container}>
      {type === 'import' && (
        <>
          <h1 className={styles.heading}>
            {t('view_ImportBackupPhrase_heading_import')}
          </h1>
          <p className={styles.info}>
            {t('view_ImportBackupPhrase_explanation_import')}
          </p>
        </>
      )}
      {type === 'reset' && (
        <>
          <h1 className={styles.heading}>
            {t('view_ImportBackupPhrase_heading_reset')}
          </h1>
          <p className={styles.info}>
            {t('view_ImportBackupPhrase_explanation_reset')}
          </p>
        </>
      )}

      <form onSubmit={handleSubmit} autoComplete="off">
        <ol className={styles.items}>
          {backupPhrase.map((word, index) => (
            <WordInput
              key={index}
              word={word}
              index={index}
              handleInput={handleInput}
              handlePaste={index === 0 ? handlePaste : undefined}
            />
          ))}
        </ol>

        <p className={styles.buttonLine}>
          <button
            type="submit"
            className={styles.button}
            disabled={Boolean(error) || backupPhrase.join('') === ''}
          >
            {t('common_action_next')}
          </button>
          <output className={styles.errorTooltip} hidden={!error}>
            {error}
          </output>
        </p>
      </form>

      <LinkBack />
    </section>
  );
}
