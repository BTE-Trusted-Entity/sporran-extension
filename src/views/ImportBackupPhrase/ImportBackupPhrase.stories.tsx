import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { ImportBackupPhrase } from './ImportBackupPhrase';

export default {
  title: 'Views/ImportBackupPhrase',
  component: ImportBackupPhrase,
} as Meta;

export function Import(): JSX.Element {
  return <ImportBackupPhrase onImport={action('onImport')} />;
}

export function Reset(): JSX.Element {
  // backup phrase: pool ahead ask clock then morning invest raise atom grace fly valve
  return (
    <ImportBackupPhrase
      onImport={action('onImport')}
      type="reset"
      address="4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
    />
  );
}
