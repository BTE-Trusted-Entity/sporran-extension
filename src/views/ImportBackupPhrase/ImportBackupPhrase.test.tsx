import userEvent from '@testing-library/user-event';
import { mnemonicValidate } from '@polkadot/util-crypto';
import { KiltKeyringPair, Utils } from '@kiltprotocol/sdk-js';

import { render, screen } from '../../testing/testing';

import { ImportBackupPhrase } from './ImportBackupPhrase';

jest.mock('@polkadot/util-crypto', () => ({
  mnemonicValidate: jest.fn(),
}));

jest.mocked(mnemonicValidate).mockReturnValue(false);

const onImport = jest.fn();

const invalidBackupPhrase = 'The entered backup phrase doesn’t exist';
const mismatchingBackupPhrase =
  'This is not the backup phrase of this Identity';
const backupPhraseNotLongEnough =
  'Please insert all 12 words of the backup phrase in the correct order';
const typo = 'It looks like there’s a typo in this word';

const props = {
  onImport,
};

async function typeElevenWords() {
  await userEvent.type(await screen.findByLabelText('1'), 'century');
  await userEvent.type(await screen.findByLabelText('2'), 'answer');
  await userEvent.type(await screen.findByLabelText('3'), 'price');
  await userEvent.type(await screen.findByLabelText('4'), 'repeat');
  await userEvent.type(await screen.findByLabelText('5'), 'carpet');
  await userEvent.type(await screen.findByLabelText('6'), 'truck');
  await userEvent.type(await screen.findByLabelText('7'), 'swarm');
  await userEvent.type(await screen.findByLabelText('8'), 'boost');
  await userEvent.type(await screen.findByLabelText('9'), 'fine');
  await userEvent.type(await screen.findByLabelText('10'), 'siege');
  await userEvent.type(await screen.findByLabelText('11'), 'brain');
}

describe('ImportBackupPhrase', () => {
  beforeEach(() => {
    jest.mocked(Utils.Crypto.makeKeypairFromUri).mockReset();
  });

  it('should render for import', async () => {
    const { container } = render(<ImportBackupPhrase {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render for reset', async () => {
    const { container } = render(
      <ImportBackupPhrase
        {...props}
        type="reset"
        address="4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should report backup word not within library', async () => {
    render(<ImportBackupPhrase {...props} />);

    await userEvent.type(await screen.findByLabelText('1'), 'oooooo');

    expect(
      await screen.findByText(typo, { selector: ':not([hidden])' }),
    ).toBeInTheDocument();
  });

  it('should ignore whitespace and be case-insensitive', async () => {
    render(<ImportBackupPhrase {...props} />);

    await userEvent.type(await screen.findByLabelText('1'), ' Century');
    await userEvent.click(await screen.findByText('Next Step'));

    expect(
      await screen.findByText(backupPhraseNotLongEnough),
    ).toBeInTheDocument();
    expect(onImport).not.toHaveBeenCalled();
  });

  it('should report backup phrase not long enough', async () => {
    render(<ImportBackupPhrase {...props} />);

    await userEvent.type(await screen.findByLabelText('1'), 'century');
    await userEvent.click(await screen.findByText('Next Step'));

    expect(
      await screen.findByText(backupPhraseNotLongEnough),
    ).toBeInTheDocument();
    expect(onImport).not.toHaveBeenCalled();
  });

  it('should report invalid backup phrase', async () => {
    render(<ImportBackupPhrase {...props} />);

    await typeElevenWords();
    await userEvent.type(await screen.findByLabelText('12'), 'brain');

    await userEvent.click(await screen.findByText('Next Step'));

    expect(await screen.findByText(invalidBackupPhrase)).toBeInTheDocument();
    expect(onImport).not.toHaveBeenCalled();
  });

  it('should report mismatching backup phrase', async () => {
    jest.mocked(mnemonicValidate).mockReturnValue(true);
    jest.mocked(Utils.Crypto.makeKeypairFromUri).mockReturnValue({
      address: 'FAIL',
    } as unknown as KiltKeyringPair);

    render(<ImportBackupPhrase {...props} type="reset" address="foo" />);

    await typeElevenWords();
    await userEvent.type(await screen.findByLabelText('12'), 'fog');

    await userEvent.click(await screen.findByText('Next Step'));

    expect(
      await screen.findByText(mismatchingBackupPhrase),
    ).toBeInTheDocument();
    expect(onImport).not.toHaveBeenCalled();
  });

  it('should allow backup phrase import', async () => {
    jest.mocked(mnemonicValidate).mockReturnValue(true);
    jest.mocked(Utils.Crypto.makeKeypairFromUri).mockReturnValue({
      address: 'PASS',
    } as unknown as KiltKeyringPair);

    render(<ImportBackupPhrase {...props} address="PASS" />);

    await typeElevenWords();
    await userEvent.type(await screen.findByLabelText('12'), 'fog');

    expect(screen.queryByText(invalidBackupPhrase)).not.toBeInTheDocument();
    expect(
      screen.queryByText(typo, { selector: ':not([hidden])' }),
    ).not.toBeInTheDocument();

    await userEvent.click(await screen.findByText('Next Step'));
    expect(onImport).toHaveBeenCalledWith(
      'century answer price repeat carpet truck swarm boost fine siege brain fog',
    );
  });

  it('should allow backup phrase reset', async () => {
    jest.mocked(mnemonicValidate).mockReturnValue(true);
    jest.mocked(Utils.Crypto.makeKeypairFromUri).mockReturnValue({
      address: '4p273cfeZ2JRz46AcJoQvTRHCH8Vaj92jts2VxepZtQwbTBB',
    } as unknown as KiltKeyringPair);

    render(
      <ImportBackupPhrase
        {...props}
        type="reset"
        address="4p273cfeZ2JRz46AcJoQvTRHCH8Vaj92jts2VxepZtQwbTBB"
      />,
    );

    await typeElevenWords();
    await userEvent.type(await screen.findByLabelText('12'), 'fog');

    expect(screen.queryByText(invalidBackupPhrase)).not.toBeInTheDocument();
    expect(screen.queryByText(mismatchingBackupPhrase)).not.toBeInTheDocument();
    expect(
      screen.queryByText(typo, { selector: ':not([hidden])' }),
    ).not.toBeInTheDocument();

    await userEvent.click(await screen.findByText('Next Step'));
    expect(onImport).toHaveBeenCalledWith(
      'century answer price repeat carpet truck swarm boost fine siege brain fog',
    );
  });
});
