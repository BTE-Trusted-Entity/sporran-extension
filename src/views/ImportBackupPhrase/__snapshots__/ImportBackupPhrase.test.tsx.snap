// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportBackupPhrase should render for import 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Import an Identity
    </h1>
    <p
      class="info"
    >
      To import a KILT Identity, you will need your backup phrase consisting of 12 words.
    </p>
    <form
      autocomplete="off"
    >
      <ol
        class="items"
      >
        <li
          class="item"
        >
          <input
            aria-label="1"
            class="input"
            id="0"
            name="0"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="0"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="2"
            class="input"
            id="1"
            name="1"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="1"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="3"
            class="input"
            id="2"
            name="2"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="2"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="4"
            class="input"
            id="3"
            name="3"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="3"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="5"
            class="input"
            id="4"
            name="4"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="4"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="6"
            class="input"
            id="5"
            name="5"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="5"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="7"
            class="input"
            id="6"
            name="6"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="6"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="8"
            class="input"
            id="7"
            name="7"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="7"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="9"
            class="input"
            id="8"
            name="8"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="8"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="10"
            class="input"
            id="9"
            name="9"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="9"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="11"
            class="input"
            id="10"
            name="10"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="10"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="12"
            class="input"
            id="11"
            name="11"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="11"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
      </ol>
      <p
        class="buttonLine"
      >
        <button
          class="button"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </section>
</div>
`;

exports[`ImportBackupPhrase should render for reset 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Set a new password
    </h1>
    <p
      class="info"
    >
      In order to set a new password you have to re-enter your backup phrase
    </p>
    <form
      autocomplete="off"
    >
      <ol
        class="items"
      >
        <li
          class="item"
        >
          <input
            aria-label="1"
            class="input"
            id="0"
            name="0"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="0"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="2"
            class="input"
            id="1"
            name="1"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="1"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="3"
            class="input"
            id="2"
            name="2"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="2"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="4"
            class="input"
            id="3"
            name="3"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="3"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="5"
            class="input"
            id="4"
            name="4"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="4"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="6"
            class="input"
            id="5"
            name="5"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="5"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="7"
            class="input"
            id="6"
            name="6"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="6"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="8"
            class="input"
            id="7"
            name="7"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="7"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="9"
            class="input"
            id="8"
            name="8"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="8"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="10"
            class="input"
            id="9"
            name="9"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="9"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="11"
            class="input"
            id="10"
            name="10"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="10"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
        <li
          class="item"
        >
          <input
            aria-label="12"
            class="input"
            id="11"
            name="11"
            type="text"
            value=""
          />
          <output
            class="wordErrorTooltip"
            for="11"
            hidden=""
          >
            It looks like there’s a typo in this word
          </output>
        </li>
      </ol>
      <p
        class="buttonLine"
      >
        <button
          class="button"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </section>
</div>
`;
